File: src/main/java/tland/gb/mem/RAM.java
package tland.gb.mem;

import java.util.Arrays;
import java.util.Random;

/**
 * Random Access Memory. Implementation of memory that can be both read and
 * written to.
 */
public class RAM implements WritableMemory, ReadableMemory {
    public final int size;
    
    protected byte[] bytes;
    private final Random rand;

    public RAM(int size) {
        this.size = size;
        bytes = new byte[size];
        rand = new Random();
    }

    @Override
    public byte readByte(int address) throws IndexOutOfBoundsException {
        return bytes[address];
    }

    @Override
    public void writeByte(int address, byte value) throws IndexOutOfBoundsException {
        bytes[address] = value;
    }

    /**
     * Clear the memory.
     * Each byte will have a value of $00.
     */
    public void clear() {
        Arrays.fill(bytes, (byte)0);
    }

    /**
     * Randomize each byte in memory.
     * Each byte will have a value from $00 to $ff.
     */
    public void randomize() {
        rand.nextBytes(bytes);
    }

}

File: core/src/main/java/iskake/jgbe/core/gb/mem/RAM.kt
package iskake.jgbe.core.gb.mem

import java.util.*

/**
 * Random Access Memory. Implementation of memory that can be both read and
 * written to.
 */
open class RAM(size: Int) : WritableMemory, ReadableMemory {
    val size = size.toUShort()
    protected var bytes: ByteArray = ByteArray(size)
    private val rand: Random = Random()

    @Throws(IndexOutOfBoundsException::class)
    override fun read(address: Int): Byte {
        return bytes[address]
    }

    @Throws(IndexOutOfBoundsException::class)
    override fun write(address: Int, value: Byte) {
        bytes[address] = value
    }

    /**
     * Clear the memory.
     * Each byte will have a value of $00.
     */
    fun clear() {
        Arrays.fill(bytes, 0.toByte())
    }

    /**
     * Randomize each byte in memory.
     * Each byte will have a value from $00 to $ff.
     */
    fun randomize() {
        rand.nextBytes(bytes)
    }
}